presto-router/src/test/java/com/facebook/presto/router/TestCustomSchedulerManager.java

```diff
@@ -0,0 +1,232 @@
+        }
+    }
+
+    public static class MockScheduler
+            implements Scheduler
+    {
+        private List<URI> candidates;
+        private int requestsMade;
+
+        @Override
+        public Optional<URI> getDestination(String user)
+        {
+            return Optional.empty();
+        }
+
+        @Override
+        public Optional<URI> getDestination(String user, String query)
+        {
+            ++requestsMade;
+            return Optional.of(candidates.get(0));
+        }
+
+        @Override
+        public void setCandidates(List<URI> candidates)
+        {
+            this.candidates = candidates;
+        }
+
+        public int getRequestsMade()
+        {
+            return requestsMade;
+        }
+    }
+
+    public static File getRouterConfigFile(List<TestingPrestoServer> servers)
+            throws IOException
+    {
+        Path path = Files.createTempFile("temp-config", ".json");
+        serverURIs = servers.stream()
+                .map(TestingPrestoServer::getBaseUrl)
+                .collect(Collectors.toList());
+        RouterSpec spec = new RouterSpec(ImmutableList.of(new GroupSpec("all", serverURIs, Optional.empty(), Optional.empty())),
+                ImmutableList.of(new SelectorRuleSpec(Optional.empty(), Optional.empty(), Optional.empty(), "all")),
+                Optional.of(CUSTOM_PLUGIN_SCHEDULER),
+                Optional.empty());
+        JsonCodec<RouterSpec> codec = jsonCodec(RouterSpec.class);
+        Files.write(path, codec.toBytes(spec));
+        return path.toFile();
+    }
+
+    private static TestingPrestoServer createPrestoServer()
+            throws Exception
+    {
+        TestingPrestoServer server = new TestingPrestoServer();
+        server.installPlugin(new TpchPlugin());
+        server.createCatalog("tpch", "tpch");
+        server.refreshNodes();
+
+        return server;
+    }
+
+    public void testQuery()
+            throws Exception
+    {
+        runAndAsserQueryResults(httpServerInfo.getHttpUri());
+
+        MockScheduler mockScheduler = (MockScheduler) schedulerManager.getScheduler();
+        assertEquals(mockScheduler.getRequestsMade(), 1);
+    }
+
+    @Test
+    public void testCustomScheduler()
+            throws Exception
+    {
+        schedulerManager.loadScheduler();
+        Scheduler scheduler = schedulerManager.getScheduler();
+        scheduler.setCandidates(serverURIs);
+
+        URI target = scheduler.getDestination("test", null).orElse(new URI("invalid"));
+        assertTrue(serverURIs.contains(target));
+    }
+
+    @DataProvider(name = "schedulerNameProvider")
+    public Object[][] schedulerNameProvider() throws URISyntaxException
+    {
+        return new Object[][]{
+                {"router-scheduler.name=RANDOM"},
+                {"router-scheduler.name=METRICS"},
+                {"router-scheduler.name=MOCK"}
+        };
+    }
+
+    @Test(dataProvider = "schedulerNameProvider")
```

Comment:
Change this to assert on the exception message -
```suggestion
@Test(dataProvider = "schedulerNameProvider",
            expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "Presto Router Scheduler.*is not registered")
    public void testCustomSchedulerFail(String schedulerName)
            throws Exception
    {
        Files.write(tempPluginSchedulerConfigFile, schedulerName.getBytes());
        schedulerManager.loadScheduler();
    }
```